Catch storage exception in scanner for remote shares

PVince81 opened the pull request:
Fixes https://github.com/owncloud/core/issues/13089

- [x] Requires https://github.com/owncloud/core/pull/13506 to be merged first

Please review @icewind1991 @DeepDiver1975 @MorrisJobke @nickvergessen @schiesbn 

Needs extensive testing, especially with sync client:

#### Test with:
- [x] remote share without password
- [x] remote share with password (this is broken for some other reason to be looked into, but should fail gracefully)
- [x] sync with remote share without password
- [x] sync with remote share with password
- [x] test expiring a remote share, folder will disappear

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/69fc048a-ff36-4a0f-828c-46d8927a9de0) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/69fc048a-ff36-4a0f-828c-46d8927a9de0) completed: **16 new issues, 2 updated code elements**

derkostka commented:
Fixes #13089 :+1: 

Warning in Log:
{"reqId":"b982fff5b9647356bc46c98a7d366886","remoteAddr":"2a02:2450:102d:1fd:219:e3ff:fe04:2514","app":"PHP","message":"Declaration of OCA\\Files_Sharing\\External\\Scanner::scanFile() should be compatible with OC\\Files\\Cache\\Scanner::scanFile($file, $reuseExisting = 0) at \/var\/www\/owncloud\/apps\/files_sharing\/lib\/external\/scanner.php#0","level":3,"time":"January 20, 2015 20:06:20","method":"GET","url":"\/index.php\/apps\/files\/ajax\/list.php?dir=%2F&sort=name&sortdirection=asc"}

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/411ac58e-2eaa-44a5-98ce-1efa4ebff467) was created.

PVince81 commented:
@derkostka thanks for spotting, fixed now.

Rebased onto master to include the required fix.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/5ddb8a04-26a2-49c9-a457-a57c67e22b6d) was created.

PVince81 commented:
- [x] test with temporarily unavailable storage, make it available again

All seems to work (apart from the password one, to be investigated here https://github.com/owncloud/core/issues/13526

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/411ac58e-2eaa-44a5-98ce-1efa4ebff467) completed: **16 new issues, 2 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/0690707c-4f19-459a-ae1a-13269da8cfec) was created.

PVince81 commented:
I added yet another change.

If the scanner did not run yet for the external share (brand new one) but curl returns "no route to host", the code throws a generic `\Exception` because we don' want to parse curl's error string for that. Anyway, in cases like this where a storage becomes unavailable, it now fails gracefully and still shows the file list. (before that last commit the spinner would stay forever)

MorrisJobke commented:
@owncloud-bot retest this please

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/5ddb8a04-26a2-49c9-a457-a57c67e22b6d) completed: **16 new issues, 2 updated code elements**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/0690707c-4f19-459a-ae1a-13269da8cfec) completed: **16 new issues, 2 updated code elements**

DeepDiver1975 commented on apps/files_sharing/lib/external/scanner.php:
```diff
@@ -18,8 +23,32 @@ public function scan($path, $recursive = self::SCAN_RECURSIVE, $reuse = -1) {
+	public function scanFile($path, $reuseExisting = 0) {
+		try {
+			return parent::scanFile($path, $reuseExisting);
+		} catch (\Sabre\DAV\Exception $e) {
```
hmm - from my understanding a sabre dav exception should be caught within the storage and be properly converted to our exceptions.

PVince81 replied:
I tried that before: need to add the exception to ALL methods in the storage then. 
The failing one was `Storage::getMimeType()`

In this specific case the exception was "AuthenticationException" which we could  convert to ForbiddenException.

Also note that this scanner implementation is kind of part of the storage :smile: (the storage returns it)

So not sure if we really need to make this change.

DeepDiver1975 commented on apps/files_sharing/lib/external/scanner.php:
```diff
@@ -18,8 +23,32 @@ public function scan($path, $recursive = self::SCAN_RECURSIVE, $reuse = -1) {
+			$data = $this->storage->getShareInfo();
+		} catch (\Exception $e) {
+			$this->storage->checkStorageAvailability();
+			throw new \Exception('Error while scanning remote share: ' . $e->getMessage());
```
can we add info on the storage or path where this fails?

PVince81 replied:
- [ ] add storage id / storage URL in exception

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/221628ff-ed93-4dbb-be5e-335101fc28de) was created.

PVince81 commented:
Fixed exceptions.

Also I added an additional fix here to catch the "wrong password" case whenever a remote share is just being added: https://github.com/owncloud/core/pull/13525/files#diff-c463069c55f33934fa9d755dff8eb13cR62

This is actually the reason for the original bug: because of https://github.com/owncloud/core/pull/13540 the password wasn't sent in the ajax request. So the remote share was imported but without the correct password, which means that initial scanning couldn't work, which caused the issue at https://github.com/owncloud/core/issues/13089

The case that was covered so far was when the remote share password changes. In such case, the scanning already worked once so this error here was never triggered.

Anyway, to make sure this doesn't happen again, now ajax/external.php is preventing to import remote shares that have the wrong password. A proper dialog can be added later in 8.1 (this PR is about failing gracefully in other cases, mostly)

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/19cb01d8-d0bb-41a6-bd8b-f57e4c08e16c) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/221628ff-ed93-4dbb-be5e-335101fc28de) completed: **20 new issues, 4 updated code elements**

PVince81 commented:
Jenkins PR here: https://github.com/owncloud/core/pull/13551

PVince81 commented:
Tests passed there: https://github.com/owncloud/core/pull/13551#issuecomment-70849437

And suprisingly Jenkins worked again here

MorrisJobke commented:
> And suprisingly Jenkins worked again here

As it is the exact same commit github is smart enough to put the green checkmark also on this PRs commit ;)

PVince81 commented:
Rebased

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/9d37097a-bcb9-4ab0-b750-f13bdad3a53d) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/9d37097a-bcb9-4ab0-b750-f13bdad3a53d) completed: **11 new issues, 4 updated code elements**

MorrisJobke commented:
@owncloud-bot retest this please

MorrisJobke commented:
@owncloud-bot retest this please

icewind1991 commented:
:+1: looks good

DeepDiver1975 commented:
> :+1: looks good

sorry - but looks good is not enough for me on this critical change - please test - THX

icewind1991 commented:
tested, all looks good

PVince81 commented:
Thanks a lot!

derkostka commented:
the fix works, i get another warning however:
{"reqId":"562fc2302cc437023e7fdd4437fdead4","remoteAddr":"2a02:2450:102d:318:219:e3ff:fe04:2514","app":"PHP","message":"Class 'OC\\Files\\View' not found at \/var\/www\/owncloud\/lib\/private\/files\/filesystem.php#298","level":3,"time":"January 22, 2015 20:02:22"}

LukasReschke commented on apps/files_sharing/lib/external/scanner.php:
```diff
@@ -18,12 +23,42 @@ public function scan($path, $recursive = self::SCAN_RECURSIVE, $reuse = -1) {
+			$this->storage->checkStorageAvailability();
+		} catch (StorageInvalidException $e) {
+			$this->storage->checkStorageAvailability();
+		} catch (StorageNotAvailable $e) {
```
Mhm… Shouldn't that be `StorageNotAvailableException`?

LukasReschke commented:
Guys, please when you review stuff please take a second to take a look at the errors pointed out by Scrutinizer. In this case it looks like 10 errors are false-positives (possibly bug in Scrutinizer) but https://github.com/owncloud/core/pull/13525#discussion_r23418126 is very much valid…

LukasReschke commented on apps/files_sharing/ajax/external.php:
```diff
@@ -46,14 +46,39 @@
+		\OCP\JSON::error(
+			array(
+				'data' => array(
+					'message' => $l->t("Could not authenticate to remote share, password might be wrong")
```
Single quotes when double quotes are not required.

PVince81 replied:
Was a copy paste :frowning:

LukasReschke commented on apps/files_sharing/ajax/external.php:
```diff
@@ -46,14 +46,39 @@
+		exit();
+	} catch (\Exception $e) {
+		$externalManager->removeShare($mount->getMountPoint());
+		\OCP\JSON::error(array('data' => array('message' => $l->t("Storage not valid"))));
```
Single quotes when double quotes are not required.

LukasReschke commented on apps/files_sharing/ajax/external.php:
```diff
@@ -46,14 +46,39 @@
+			$storage->getScanner()->scanAll();
+			\OCP\JSON::success();
+		} catch (\OCP\Files\StorageInvalidException $e) {
+			\OCP\JSON::error(array('data' => array('message' => $l->t("Storage not valid"))));
```
Single quotes when double quotes are not required.

LukasReschke commented on apps/files_sharing/ajax/external.php:
```diff
@@ -46,14 +46,39 @@
+		} catch (\OCP\Files\StorageInvalidException $e) {
+			\OCP\JSON::error(array('data' => array('message' => $l->t("Storage not valid"))));
+		} catch (\Exception $e) {
+			\OCP\JSON::error(array('data' => array('message' => $l->t("Couldn't add remote share"))));
```
Single quotes when double quotes are not required.

LukasReschke replied:
Maybe it would make sense as well to log the error somewhere?

nickvergessen replied:
Would have to use the utf8 ´ then, or escape the single quote

LukasReschke commented on apps/files_sharing/lib/external/scanner.php:
```diff
@@ -18,12 +23,42 @@ public function scan($path, $recursive = self::SCAN_RECURSIVE, $reuse = -1) {
+			throw new \Exception(
+				'Error while scanning remote share: "' .
+				$this->storage->getRemote() . '" ' .
+				$e->getMessage()
```
Where is `$e` defined?

LukasReschke replied:
Proper IDE setup helps:

![screen shot 2015-01-23 at 00 02 38](https://cloud.githubusercontent.com/assets/878997/5866604/33ec9ede-a293-11e4-8bcb-90ec60e5fa19.png)

LukasReschke commented on apps/files_sharing/lib/external/storage.php:
```diff
@@ -142,27 +142,47 @@ public function hasUpdated($path, $time) {
+				throw new StorageInvalidException();
+			} else {
+				// ownCloud instance is gone, likely to be a temporary server configuration error
 				throw $e;
```
![screen shot 2015-01-23 at 00 04 52](https://cloud.githubusercontent.com/assets/878997/5866630/78f2191e-a293-11e4-8778-c809358e19b3.png)

LukasReschke commented on lib/private/files/storage/dav.php:
```diff
@@ -529,8 +548,27 @@ public function hasUpdated($path, $time) {
+	 * @throws StorageNotAvailableException if the storage is not available,
+	 * which might be temporary
+	 */
+	private function convertSabreException(\Sabre\Dav\Exception $e) {
```
Add some unit tests for this? You can use `\Test_Helper::invokePrivate` for that.

LukasReschke commented on lib/private/files/view.php:
```diff
@@ -1037,7 +1037,17 @@ public function getDirectoryContent($directory, $mimetype_filter = '') {
+							continue;
+						} catch (\Exception $e) {
+							// sometimes when the storage is not available it can be any exception
+							\OC_Log::write('core', 'Exception while scanning storage "' . $subStorage->getId() . '": ' . $e->getMessage(), \OC_Log::ERROR);
```
![screen shot 2015-01-23 at 00 08 34](https://cloud.githubusercontent.com/assets/878997/5866688/fef37012-a293-11e4-9251-472037b41c82.png)

LukasReschke commented:
:-1: requires some adjustments. – Didn't test functionality. Pure code review.

LukasReschke commented on apps/files_sharing/lib/external/scanner.php:
```diff
@@ -18,12 +23,42 @@ public function scan($path, $recursive = self::SCAN_RECURSIVE, $reuse = -1) {
 		$this->scanAll();
 	}
 
+	public function scanFile($path, $reuseExisting = 0) {
```
PHPDoc please

LukasReschke commented on apps/files_sharing/lib/external/scanner.php:
```diff
@@ -18,12 +23,42 @@ public function scan($path, $recursive = self::SCAN_RECURSIVE, $reuse = -1) {
+		}
+	}
+
 	public function scanAll() {
```
When already touching this please add PHPDoc as well

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/39db76b3-9452-4a62-bedf-dc7980dfde57) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/a168fcf6-8fc3-4469-a27a-e11b65ed7084) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/39db76b3-9452-4a62-bedf-dc7980dfde57) completed: **19 new issues, 4 updated code elements**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/a168fcf6-8fc3-4469-a27a-e11b65ed7084) completed: **19 new issues, 4 updated code elements**

PVince81 commented:
Fixes, please re-review.

PVince81 commented:
I'll make another Jenkins PR

derkostka commented:
I re-tested your changes. OSX Client also works fine for me (did some up/downloads yesterday using txt files and also zip archives). :+1:

PVince81 commented:
Thanks a lot @derkostka for your help all along :smile: 

Jenkins PR here: https://github.com/owncloud/core/pull/13664

PVince81 commented:
Tests passed, merge ?

nickvergessen commented on apps/files_sharing/ajax/external.php:
```diff
@@ -42,20 +42,71 @@
+		);
+		exit();
+	} catch (\Exception $e) {
+		throw new StorageNotAvailableException(get_class($e).": ".$e->getMessage());
```
missing the `\` ?

nickvergessen replied:
This should be `\OCP\Files\StorageNotAvailableException`

So if tests pass, it seems to miss a test case ;)

PVince81 replied:
There are no unit tests for that :frowning:

nickvergessen replied:
Well see the image below, some IDEs report that the class does not exist ;)

nickvergessen commented on apps/files_sharing/ajax/external.php:
```diff
@@ -42,20 +42,71 @@
+		exit();
+	} catch (\Exception $e) {
+		throw new StorageNotAvailableException(get_class($e).": ".$e->getMessage());
+		\OCP\Util::writeLog(
```
Log is never written

nickvergessen replied:
![unreachable](https://cloud.githubusercontent.com/assets/213943/5898853/f495792c-a553-11e4-873a-dbd5116c19a0.png)

nickvergessen commented on apps/files_sharing/ajax/external.php:
```diff
@@ -42,20 +42,71 @@
+		);
+		$externalManager->removeShare($mount->getMountPoint());
+		\OCP\JSON::error(array('data' => array('message' => $l->t('Storage not valid'))));
+		exit();
```
since the `throw` above, everything afterwards can be removed

nickvergessen commented on lib/private/files/view.php:
```diff
@@ -1037,7 +1037,22 @@ public function getDirectoryContent($directory, $mimetype_filter = '') {
+								'core',
+								'Exception while scanning storage "' . $subStorage->getId() . '": ' .
+								get_class($e) . ': ' . $e->getMessage(),
+								\OC_Log::ERROR
```
`\OCP\Util::ERROR`

nickvergessen commented:
Looks good, apart from:
- [ ] broken exception name
- [ ] unreachable code
- [ ] use of deprecated OC_Log

poke again when fixed, for a +1

PVince81 commented:
- [ ] try adding unit tests, at least it would get these various code paths covered

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/4313e915-7fb0-4078-a9c4-844f54bfbd04) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/4313e915-7fb0-4078-a9c4-844f54bfbd04) completed: **14 new issues, 4 updated code elements**

PVince81 commented:
@nickvergessen please have another look. Thanks.

I'll retrigger Jenkins separately.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/5d46489e-eb34-407e-b5c5-bd7a19e4f56f) was created.

nickvergessen commented on apps/files_sharing/ajax/external.php:
```diff
@@ -42,20 +42,70 @@
+		);
+		$externalManager->removeShare($mount->getMountPoint());
+		\OCP\JSON::error(array('data' => array('message' => $l->t('Storage not valid'))));
+		throw new \OCP\Files\StorageNotAvailableException(get_class($e).': '.$e->getMessage());
```
magicians need to tell me whether outputting something + throwing an exception is the right thing to do.
the previous catch only does JSON::error() and then exits, which I think should be done here aswell.

PVince81 replied:
I wasn't sure either. I thought that this file might be called somehow from base.php so throwing the exception will... hmmmm right, it will return an error page, which we don't want.

I'll add back `exit()`.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/be9fb1bf-054e-4481-925c-58bbbee6d03f) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/5d46489e-eb34-407e-b5c5-bd7a19e4f56f) completed: **14 new issues, 4 updated code elements**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/be9fb1bf-054e-4481-925c-58bbbee6d03f) completed: **12 new issues, 4 updated code elements**

LukasReschke commented:
:+1:

PVince81 commented:
@nickvergessen @MorrisJobke second review ? (there were a few fixes since last time)

MorrisJobke commented:
testing ... ;)

MorrisJobke commented:
Works :+1: (all files are synced now and work)

MorrisJobke commented:
Jenkins PR #13675

PVince81 commented:
Another Jenkins PR, hopefully the last one: https://github.com/owncloud/core/pull/13714

MorrisJobke commented:
https://github.com/owncloud/core/pull/13714#issuecomment-71697160 <- success
